active/0000-addon-tree-caching.md

```diff
@@ -0,0 +1,69 @@
+# Ember CLI Addon Deduplication
+
+- Start Date: 2016-12-11
+- RFC PR: (leave this empty)
+- Ember CLI Issue: (leave this empty)
+ 
+# Summary
+
+In Ember CLI today, all addons at each level are built through the standard `treeFor` / `treeFor*` hooks. These hooks are responsible for preprocessing the JavaScript included by the tree returned from that specific hook (e.g., `treeForAddon` preprocesses the JS for the addon tree). This RFC proposes a mechanism that would allow these returned trees to be cached by default (when no build time customization is done) and expose proper hooks for addon authors to control the degree to which we dedupe these trees.
+
+# Motivation
+
+Today, given the dependency graph:
+
+``` 
+ember-basic-dropdown:
+  ember-wormhole@0.4.1
+
+ember-modal-dialog:
+  ember-wormhole@0.4.1
+
+ember-paper:
+  ember-wormhole@0.4.1
+```
+We would actually build `ember-wormhole`'s `addon` tree 3 different times, even though [as you can see](https://github.com/yapplabs/ember-wormhole/blob/0.4.1/index.js) there is absolutely no build time customization being done. After all of these `ember-wormhole` tree instances are built, we merge them such that the last tree wins (thus making all of the work to preprocess these trees completely moot). If you extrapolate this out to larger applications or ones using multiple engines (lazy or not) it is fairly common to see these sorts of dependencies shared upwards of 4 to 5 times. This can lead to significant build performance degradation.
+
+# Detailed design
+
+- Add a `Addon.prototype.cacheKeyForTree` method to [lib/models/addon.js](https://github.com/ember-cli/ember-cli/commits/master/lib/models/addon.js) that is invoked prior to calling `treeFor` for the same tree name. The `Addon.prototype.cacheKeyForTree` method is expected to return a cache key allowing multiple builds of the same tree to simply return the original tree (preventing duplicate work). If `Addon.prototype.cacheKeyForTree` returns `null` / `undefined` the tree in question will opt out of this caching system.
+- ember-cli's custom [`mergeTrees` implementation](https://github.com/ember-cli/ember-cli/blob/4ec7b5951e8a9dd292029faf20d1858abf7bdfa0/lib/broccoli/merge-trees.js) (which is already aware of other tree reduction techniques) will be updated so that calling `mergeTrees([treeA, treeA]);` simply returns `treeA`, and `mergeTrees([treeA, treeB, treeA])` removes the duplicated `treeA` in the input nodes.
+ 
+The proposed declaration for `Addon.prototype.cacheKeyForTree` in Typescript syntax is:
+
+``` ts
+function cacheKeyForTree(treeType: string): string;
+```
+The default implementation for `Addon.prototype.cacheKeyForTree` will:
+
+- Utilize a shared NPM package (e.g. `calculate-cache-key-for-tree`) that will generate a cache key that incorporates at least the following pieces of information:
```

Comment:
An additional or complementary heuristic may be that they are at the exact same place on disk?